internal/mkcgo/mkcgo_test.go

```diff
@@ -0,0 +1,402 @@
+					{Name: "F1", Ret: "int", Params: []*mkcgo.Param{{"int", "p1"}}, FuncAttrs: mkcgo.FuncAttrs{ErrCond: "ec0", NoEscape: true, NoCallback: true}},
+					{Name: "F2", Ret: "int*", Params: []*mkcgo.Param{{"int**", "p1"}, {"void*", "p2"}}},
+					{Name: "F3", Ret: "int*", Params: []*mkcgo.Param{{"int", "p1"}, {"void***", ""}}, FuncAttrs: mkcgo.FuncAttrs{Optional: true}},
+					{Name: "F4", Ret: "unsigned long", Params: []*mkcgo.Param{{"int", "__func"}, {"int", "__type"}, {"...", ""}}},
+					{Name: "F5", Ret: "int*", Params: []*mkcgo.Param{{"float", ""}, {"double", ""}}, FuncAttrs: mkcgo.FuncAttrs{VariadicTarget: "F4"}},
+					{Name: "F6", Ret: "void", Params: []*mkcgo.Param{{"void", ""}}},
+				},
+			},
+		},
+	}
+	for i, tt := range tests {
+		// No need to specify a test name, the error message is enough to identify the test.
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			var got mkcgo.Source
+			if err := got.Parse(strings.NewReader(tt.content)); err != nil {
+				t.Fatal(err)
+			}
+			testSlice(t, "Includes", got.Includes, tt.want.Includes)
+			testSlice(t, "Comments", got.Comments, tt.want.Comments)
+			testSlice(t, "TypeDefs", got.TypeDefs, tt.want.TypeDefs)
+			testSlice(t, "Enums", got.Enums, tt.want.Enums)
+			testSlice(t, "Funcs", got.Funcs, tt.want.Funcs)
+		})
+	}
+}
+
+func TestParseError(t *testing.T) {
+	tests := []struct {
+		content string
+		want    string
+	}{
+		{
+			content: `/**/`,
+			want:    `block comments are not supported`,
+		}, {
+			content: "enum {\nE0 1,\n};",
+			want:    `can't parse enum in line "E0 1,": can't extract enum value`,
+		}, {
+			content: `typedef T0;`,
+			want:    `can't extract type name`,
+		}, {
+			content: `void foo(;`,
+			want:    `invalid signature`,
+		}, {
+			content: `void;`,
+			want:    `invalid signature`,
+		}, {
+			content: `void (void);`,
+			want:    `missing name`,
+		}, {
+			content: `void foo(,);`,
+			want:    `empty parameter`,
+		}, {
+			content: `void foo2(int a) __attribute__((variadic("foo")));`,
+			want:    "variadic target not added yet",
+		}, {
+			content: `
+void foo();
+void foo2(int a) __attribute__((variadic("foo")));
+`, want: "variadic target is not variadic",
+		}, {
+			content: `void foo(void) __attribute__();`,
+			want:    `can't extract function attributes: __attribute__ should be followed by double parentheses`,
+		}, {
+			content: `void foo(void) __attribute__;`,
+			want:    `can't extract function attributes: __attribute__ should be followed by double parentheses`,
+		}, {
+			content: `void foo(void) __attribute__((tag("a")optional));`,
+			want:    `can't extract function attributes: parameters must be separated by commas`,
+		}, {
+			content: `void foo(void) __attribute__((tag("a"());`,
+			want:    `can't extract function attributes: unbalanced parentheses`,
+		}, {
+			content: `void foo(void) __attribute__(());`,
+			want:    `can't extract function attributes: empty attribute`,
```

Comment:
An empty attribute list is valid, according to https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html.